lib/src/test/java/io/ably/lib/transport/HostsTest.java

```diff
@@ -0,0 +1,229 @@
+        String host = "overridden.ably.io";
+        Hosts hosts = new Hosts(host, Defaults.HOST_REALTIME, options);
+        int idx;
+        boolean shuffled = false;
+        boolean allFound = true;
+        for (idx = 0; ; ++idx) {
+            host = hosts.getFallback(host);
+            if (host == null)
+                break;
+            int found = Arrays.asList(Defaults.HOST_FALLBACKS).indexOf(host);
+            if (found < 0)
+                allFound = false;
+            else if (found != idx)
+                shuffled = true;
+        }
+        assertThat(idx, is(Defaults.HOST_FALLBACKS.length));
+        assertTrue(allFound);
+        assertTrue(shuffled);
+    }
+
+    /**
+     * It is not allowed to use default fallback hosts and at the same time
+     * provide custom fallback hosts.
+     */
+    @Test(expected = AblyException.class)
+    public void hosts_fallback_use_default_and_set_fallback_hosts() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        options.fallbackHostsUseDefault = true;
+        options.fallbackHosts = new String[] { "custom.ably-realtime.com" };
+
+        new Hosts(null, Defaults.HOST_REALTIME, options);
+    }
+
+    /**
+     * Expect that returned fallback hosts containing the environment information.
+     */
+    @Test
+    public void hosts_fallback_use_environment() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        options.environment = "sandbox";
+        String[] expectedEnvironmentFallbackHosts = Defaults.getEnvironmentFallbackHosts(options.environment);
+        Hosts hosts = new Hosts(null, Defaults.HOST_REALTIME, options);
+
+        String host = "sandbox-" + Defaults.HOST_REALTIME;
+        List<String> returnedEnvironmentFallbackHosts = new ArrayList<>();
+        while ((host = hosts.getFallback(host)) != null){
+            returnedEnvironmentFallbackHosts.add(host);
+        }
+
+        assertThat(returnedEnvironmentFallbackHosts, containsInAnyOrder(expectedEnvironmentFallbackHosts));
+    }
+
+    /**
+     * Expect that returned default fallback hosts without environment according to RSC15g4.
+     */
+    @Test
+    public void hosts_fallback_use_default_fallback_hosts_and_environment() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        options.fallbackHostsUseDefault = true;
+        options.environment = "sandbox";
+        Hosts hosts = new Hosts(null, Defaults.HOST_REALTIME, options);
+        String host = "sandbox-" + Defaults.HOST_REALTIME;
+        List<String> returnedEnvironmentFallbackHosts = new ArrayList<>();
+
+        while ((host = hosts.getFallback(host)) != null){
+            returnedEnvironmentFallbackHosts.add(host);
+        }
+
+        assertThat(returnedEnvironmentFallbackHosts, containsInAnyOrder(Defaults.HOST_FALLBACKS));
+    }
+
+    @Test
+    public void hosts_no_fallback_when_port_is_defined() throws AblyException {
```

Comment:
Can we add more tests for ports, 
1. It should return fallback when custom fallbacks are provided. 
2. If the host is provided along with fallbacks, it should return fallbacks.